### PR TITLE
Octal ints

### PR DESCRIPTION
**Pull request by jltallon**
Support for converting values expressed in octal. Range limited to INT_MIN..0..INT_MAX. 
Who would want to write a long long in octal?

Intended to be used for umasks, UN*X file permissions and the like.

The only potential incompatibility is that it sets -std=c99 (for full support of strtoull)... but all supported compilers should support that anyway. Would benefit from a subsequent patch that'll remove the newly emited warnings (not in this code), though

### CODE REVIEW

File: lib/scanner.l [hunk 1]
```diff
@@ -167,16 +168,31 @@ include_open      ^[ \t]*@include[ \t]+\"
...
+                    int errsave = errno;
+                    errno = 0;
+                    llval = strtoll(yytext, &endptr, 0);	/* base 10 or base 8 */
+                    if(*endptr || errno)
+                    {
+                      errno = 0;
+                      return(TOK_ERROR);	/* some error occured ... */
+                    }
+                    errno = errsave;
+                    if((*yytext=='0') && (*(yytext+1)!='\0'))
```
Comment (hyperrealm) [inline lib/scanner.l]: add spaces around operators

File: lib/scanner.l [hunk 2]
```diff
@@ -167,16 +168,31 @@ include_open      ^[ \t]*@include[ \t]+\"
...
+                    {   /* it's octal... so INT we go */
+                      yylval->ival=(int)(llval);
```
Comment (hyperrealm) [inline lib/scanner.l]: add spaces around =

Comment (jltallon) [inline lib/scanner.l]: On 12/29/2015 11:41 PM, hyperrealm wrote:

Done.
New commit: 
https://github.com/jltallon/libconfig/commit/5db0a60a5e907bf7afa7ebb95b59d30551c30453


     Just tell me if you prefer me to rebase this or the other one... or 
neither.

File: lib/scanner.l [hunk 3]
```diff
@@ -167,16 +168,31 @@ include_open      ^[ \t]*@include[ \t]+\"
...
+                    int errsave = errno;
+                    errno = 0;
+                    llval = strtoll(yytext, &endptr, 0);	/* base 10 or base 8 */
+                    if(*endptr || errno)
+                    {
+                      errno = 0;
+                      return(TOK_ERROR);	/* some error occured ... */
+                    }
+                    errno = errsave;
+                    if((*yytext == '0') && (*(yytext+1) != '\0'))
```
Comment (hyperrealm) [inline lib/scanner.l]: One more thing occurred to me last night about this change. I think it'd be simpler to declare a separate token for octals:

octal    0[1-7][0-7]*

Then all this error checking logic here wouldn't be necessary; you'd simply have a rule:

{octal}   { yylval->llval = strtoul(yytext, &endptr, 8); return(TOK_OCTAL); }

Very similar to the rule for {hex}.

We could also add support for binary literals in similar fashion.

{binary}   0b[0-1]+

**Comment by jltallon** [comment]
As you prefer. This approach minimized the code changes.

Regarding binary: this is *config parsing* library, not a general-purpose parse-all library. Who uses binary literals in configuration files??
- Integers: all the time
- Octals: for UN*X permissions (quite frequently)
- Hex: for bitmasks and special values
- Binary literals converted to integers??? We could then support base64-encoded BLOBs ....

 ...or directly, parse JSON :S

**Pull request merged**